File: detection_ws/src/value_map/value_map/service_handler.py
from seem_ros_interfaces.srv import Panoptic, ObjectSegmentation, SemanticSimilarity
import rclpy
from cv_bridge import CvBridge
import cv2

#from vlm_interface.srv import SemanticSimilarity

# ANSI color codes
BLUE = "\033[94m"
GREEN = "\033[92m"
YELLOW = "\033[93m"
RED = "\033[91m"
BOLD = "\033[1m"
RESET = "\033[0m"

class ServiceHandler:
    def __init__(self, node):
        self.node = node
        self.bridge = CvBridge()

        self.node.declare_parameter('vlm_target_namespace', '/seem_ros')
        self.vlm_namespace = self.node.get_parameter('vlm_target_namespace').get_parameter_value().string_value.rstrip('/')

        def resolve(service: str) -> str:
            return f'{self.vlm_namespace}/{service}'

        # Create clients
        self.cli_panoptic = self.node.create_client(Panoptic, resolve('panoptic_segmentation'))
        self.cli_object_segmentation = self.node.create_client(ObjectSegmentation, resolve('object_segmentation'))
        self.cli_semantic_similarity = self.node.create_client(SemanticSimilarity, resolve('semantic_similarity'))

        self.node.get_logger().info(
            f"{BLUE}{BOLD}ServiceHandler initialized with VLM namespace: {self.vlm_namespace}{RESET}"
        )

        self.panoptic_segemented_image = None
        self.object_segmented_image = None
        self.semantic_similarity_score = None

    def call_panoptic(self, image_msg):
        request = Panoptic.Request()
        request.image = image_msg
        future = self.cli_panoptic.call_async(request)
        future.add_done_callback(self.handle_panoptic_response)
        return self.panoptic_segemented_image

    def call_object_segmentation(self, image_msg, query):
        request = ObjectSegmentation.Request()
        request.image = image_msg
        request.query = query
        future = self.cli_object_segmentation.call_async(request)
        future.add_done_callback(self.handle_object_response)
        return self.object_segmented_image

    def call_semantic_similarity(self, image_msg, query):
        request = SemanticSimilarity.Request()
        request.image = image_msg
        request.query = query
        future = self.cli_semantic_similarity.call_async(request)
        future.add_done_callback(self.handle_similarity_response)
        return self.semantic_similarity_score

    def handle_panoptic_response(self, future):
        try:
            response = future.result()
            self.panoptic_segemented_image = response.panoptic_segmentation
            image = self.bridge.imgmsg_to_cv2(response.panoptic_segmentation, desired_encoding='rgb8')
            cv2.imshow("Panoptic Segmentation", image)
            cv2.waitKey(1)
        except Exception as e:
            self.panoptic_segemented_image = None
            self.node.get_logger().error(f"Panoptic response error: {e}")

    def handle_object_response(self, future):
        try:
            response = future.result()
            self.object_segmented_image = response.segmented_image
            image = self.bridge.imgmsg_to_cv2(response.segmented_image, desired_encoding='rgb8')
            cv2.imshow("Object Segmentation", image)
            cv2.waitKey(1)
        except Exception as e:
            self.object_segmented_image = None
            self.node.get_logger().error(f"Object response error: {e}")

    def handle_similarity_response(self, future):
        try:
            response = future.result()
            self.semantic_similarity_score = response.score
        except Exception as e:
            self.semantic_similarity_score = None
            self.node.get_logger().error(f"Similarity response error: {e}")

File: detection_ws/src/value_map/src/value_map/service_handler.cpp
#include "service_handler.hpp"

ServiceHandler::ServiceHandler(rclcpp_lifecycle::LifecycleNode* node) : node(node)
{

}

ServiceHandler::~ServiceHandler()
{

}

bool ServiceHandler::on_configure()
{
    RCLCPP_INFO(node->get_logger(), "Configuring ServiceHandler...");

    vlmNamespace = node->get_parameter("vlm_target_namespace").as_string();
    if (!vlmNamespace.empty() && vlmNamespace.back() == '/')
        vlmNamespace.pop_back();

    auto resolve = [this](const std::string &srv) {
        return vlmNamespace + "/" + srv;
    };

    RCLCPP_INFO(node->get_logger(), "vlm_target_namespace: %s", vlmNamespace.c_str());

    // cliPanoptic = node->create_client<seem_ros_interfaces::srv::Panoptic>(resolve("panoptic_segmentation"));
    // cliObjSeg = node->create_client<seem_ros_interfaces::srv::ObjectSegmentation>(resolve("object_segmentation"));
    // cliSemSim = node->create_client<seem_ros_interfaces::srv::SemanticSimilarity>(resolve("semantic_similarity"));
    cliSemSim = node->create_client<vlm_interface::srv::SemanticSimilarity>(resolve("semantic_similarity"));

    RCLCPP_INFO(node->get_logger(), "ServiceHandler configured successfully.");

    return true;
}

bool ServiceHandler::on_activate()
{
    RCLCPP_INFO(node->get_logger(), "Activating ServiceHandler...");

    
    RCLCPP_INFO(node->get_logger(), "ServiceHandler activated successfully.");
    return true;
}

bool ServiceHandler::on_deactivate()
{
    RCLCPP_INFO(node->get_logger(), "Deactivating ServiceHandler...");

    RCLCPP_INFO(node->get_logger(), "ServiceHandler deactivated successfully.");
    return true;
}

bool ServiceHandler::on_cleanup()
{
    RCLCPP_INFO(node->get_logger(), "Cleaning up ServiceHandler...");

    RCLCPP_INFO(node->get_logger(), "ServiceHandler cleaned up successfully.");
    return true;
}

bool ServiceHandler::on_shutdown()
{
    RCLCPP_INFO(node->get_logger(), "Shutting down ServiceHandler...");

    RCLCPP_INFO(node->get_logger(), "ServiceHandler shut down successfully.");
    return true;
}

// void ServiceHandler::callPanoptic(const sensor_msgs::msg::Image &image_msg)
// {
//     auto req = std::make_shared<seem_ros_interfaces::srv::Panoptic::Request>();
//     req->image = image_msg;

//     cliPanoptic->async_send_request(req,
//         [this](rclcpp::Client<seem_ros_interfaces::srv::Panoptic>::SharedFuture future) {
//             try {
//                 auto res = future.get();
//                 panopticSegmentedImage = res->panoptic_segmentation;
//                 RCLCPP_INFO(node->get_logger(), "Received panoptic segmentation response: %dx%d", 
//                            panopticSegmentedImage.width, panopticSegmentedImage.height);
//             } catch (const std::exception &e) {
//                 RCLCPP_ERROR(node->get_logger(), "Panoptic response error: %s", e.what());
//             }
//         });
// }

// void ServiceHandler::callObjectSegmentation(const sensor_msgs::msg::Image &image_msg, const std::string &query)
// {
//     auto req = std::make_shared<seem_ros_interfaces::srv::ObjectSegmentation::Request>();
//     req->image = image_msg;
//     req->query = query;

//     cliObjSeg->async_send_request(req,
//         [this](rclcpp::Client<seem_ros_interfaces::srv::ObjectSegmentation>::SharedFuture future) {
//             try {
//                 auto res = future.get();
//                 objectSegmentedImage = res->segmented_image;
//                 RCLCPP_INFO(node->get_logger(), "Received object segmentation response: %dx%d", 
//                            objectSegmentedImage.width, objectSegmentedImage.height);
//             } catch (const std::exception &e) {
//                 RCLCPP_ERROR(node->get_logger(), "ObjectSegmentation response error: %s", e.what());
//             }
//         });
// }

void ServiceHandler::callSemanticSimilarity(const sensor_msgs::msg::Image &image_msg, const std::string &query)
{
    RCLCPP_DEBUG(node->get_logger(), "Starting SemanticSimilarity service call...");
    RCLCPP_DEBUG(node->get_logger(), "Image: %dx%d, encoding: %s, query: '%s'", 
                image_msg.width, image_msg.height, image_msg.encoding.c_str(), query.c_str());

    // Check if service is available
    if (!cliSemSim->service_is_ready()) 
    {
        RCLCPP_WARN(node->get_logger(), "SemanticSimilarity service is not ready, waiting...");
        
        if (!cliSemSim->wait_for_service(std::chrono::seconds(5))) {
            RCLCPP_ERROR(node->get_logger(), "SemanticSimilarity service not available after 5 seconds");
            return;
        }
        
        RCLCPP_INFO(node->get_logger(), "SemanticSimilarity service is now available");
    } else 
    {
        RCLCPP_DEBUG(node->get_logger(), "SemanticSimilarity service is ready");
    }

    auto req = std::make_shared<vlm_interface::srv::SemanticSimilarity::Request>();
    req->image = image_msg;
    req->query = query;

    RCLCPP_DEBUG(node->get_logger(), "Sending async request to SemanticSimilarity service...");

    cliSemSim->async_send_request(req,
        [this, query](rclcpp::Client<vlm_interface::srv::SemanticSimilarity>::SharedFuture future) {
            RCLCPP_DEBUG(node->get_logger(), "Received response from SemanticSimilarity service");

            try {
                auto res = future.get();
                semanticSimilarityScore = res->score;
                
                RCLCPP_DEBUG(node->get_logger(), "SemanticSimilarity SUCCESS: query='%s' -> score=%.4f", 
                           query.c_str(), semanticSimilarityScore);
                
            } catch (const std::exception &e) {
                RCLCPP_ERROR(node->get_logger(), "SemanticSimilarity response error: %s", e.what());
                semanticSimilarityScore = 0.0; // Set default value on error
                
            } catch (...) {
                RCLCPP_ERROR(node->get_logger(), "SemanticSimilarity unknown error occurred");
                semanticSimilarityScore = 0.0;
            }
        });

    RCLCPP_DEBUG(node->get_logger(), "Async request sent, waiting for response...");
}

// sensor_msgs::msg::Image ServiceHandler::getPanopticSegmentedImage(const sensor_msgs::msg::Image &image_msg)
// {
//     callPanoptic(image_msg);
//     return panopticSegmentedImage;
// }

// sensor_msgs::msg::Image ServiceHandler::getObjectSegmentedImage(const sensor_msgs::msg::Image &image_msg, const std::string &query)
// {
//     callObjectSegmentation(image_msg, query);
//     return objectSegmentedImage;
// }

double ServiceHandler::getSemanticSimilarityScore(const sensor_msgs::msg::Image &image_msg, const std::string &query)
{
    callSemanticSimilarity(image_msg, query);
    return semanticSimilarityScore;
}